CONTRIBUTING.md

```diff
@@ -1,20 +1,52 @@
+Please make sure `make check_format` passes on your commits.  If it reports any issues, you can run `make format` to auto-apply the project formatting rules.  Note that this will format your files in-place with no built-in undo, so you may want to create a temporary commit if you have any unstaged changes.
 
-### Projects
+Adding a pre-commit hook can be useful to catch formatting errors earlier. i.e. have in `~/pyston/.git/hooks/pre-commit`:
 
-If you don't know where to start:
-- Check out our [Github issues list](https://github.com/dropbox/pyston/issues), especially those marked ["probably easy"](https://github.com/dropbox/pyston/labels/probably%20easy)
+```
+#!/bin/sh
+exec make -C ~/pyston check_format
+```
+
+### Getting started with the codebase
+
+The easiest way to contribute to Pyston is to help us improve our compatibility with CPython. There are many small tasks to do such as built-in functions that are not yet implemented, edge cases that are not being handled by Pyston or where our output is slightly different than CPython, etc. The fix will often involve a local change, giving a smooth start to learning the codebase. One of Python's greatest strengths is that it comes ["batteries included"](https://xkcd.com/353/), but this means that there is a long long tail of these little tasks that needs to be driven through - your help is immensely valuable there!
+
+The command `make quick_check` will first run our Pyston tests (great way to make sure everything is in order) and then the default CPython tests. You will get an output that looks like this:
+
+```
+                  ...
+                  test_bastion.py    Correct output (125.7ms)
+                 test_unittest.py    (skipped: expected to fail)
+                     test_json.py    (skipped: expected to fail)
+                  test_future3.py    Correct output (952.8ms)
+                  ...
+```
+
+Notice that a large number tests are currently marked as failing (marked with an `# expected: fail` comment at the top of the file). Just pick any that you think is interesting and get it to pass! Remove the `#expected: fail` flag and run `make check_TESTNAME` (without `.py`) to compare the result to CPython's (the command will search for TESTNAME in the `test/` directory). If the test is crashing, the easiest way to start debugging is to use `make dbg_TESTNAME` which is essentially `make check_TESTNAME` inside GDB.
+
+This kind of work will often happen where native libraries are defined (e.g. `src/runtime/builtin_modules/builtins.cpp`), implementation of types (e.g. `src/runtime/str.cpp`) and debugging may involve tracing through the interpreter (`src/codegen/ast_interpreter.cpp`). The code in those files should be relatively straightforward. Code that involve the JIT (function rewriting, assembly generation, etc) can be more intricate (e.g. `src/asm_writing/rewriter.cpp`). If you start running into that area of the code, it's perfectly fine to ask for help!
```

Comment:
I'd rewrite "If you start running into that area of the code, it's perfectly fine to ask for help!" to something like "Keep in mind, it's perfectly fine to ask for help!" - it's always okay to ask for help, regardless of what part of the code you're in :)